### PR TITLE
Ensure base translations are never deleted on pull command.

### PR DESCRIPTION
**Pull request by BernardoSilva**
This fixes https://github.com/partnermarketing/PartnermarketingTranslationBundle/issues/28

- [x] Add tests to prove the bug when pulling translations.
- [x] Improve code quality.

### CODE REVIEW

File: src/Utilities/HasUtilitiesTrait.php [hunk 1]
```diff
@@ -21,4 +21,23 @@ function ksortMultiDimensional(array $input) {
             }
         }
     }
+
+    function arrayMergeMultiDimensional(array &$array1, array &$array2)
```
Comment (osavchenko) [inline src/Utilities/HasUtilitiesTrait.php]: Why you pass arrays by reference?

File: src/Utilities/HasUtilitiesTrait.php [hunk 2]
```diff
@@ -21,4 +21,31 @@ function ksortMultiDimensional(array $input) {
...
+    function arrayMergeMultiDimensional(array $array1, array $array2)
+    {
+        $merged = $array1;
+
+        foreach ($array2 as $key => &$value)
+        {
+            if (is_array($value) && isset($merged[$key]) && is_array($merged[$key])) {
+                $merged[$key] = $this->arrayMergeMultiDimensional($merged[$key], $value);
+            } else if (is_numeric($key)) {
+                if (!in_array($value, $merged))
```
Comment (amyboyd) [inline src/Utilities/HasUtilitiesTrait.php]: Missing brackets here.

File: src/Utilities/HasUtilitiesTrait.php [hunk 3]
```diff
@@ -21,4 +21,31 @@ function ksortMultiDimensional(array $input) {
...
+                    $merged[] = $value;
+            } else {
+                $merged[$key] = $value;
+            }
+        }
+
+        return $merged;
+    }
```
Comment (antonvas) [inline src/Utilities/HasUtilitiesTrait.php]: Why `array_merge_recursive` is not work?

Comment (BernardoSilva) [inline src/Utilities/HasUtilitiesTrait.php]: for example:
```php
<?php
$sourceArray = [
	'page_one' => [
		'field_1' => 'field 1',
		'field_2' => [
			'sub_1' => 'sub 1',
			'sub_2' => 'sub 2'
		]
	]
];

$remoteArray = [
	'page_one' => [
		'field_2' => [
			'sub_2' => 'sub 2'
		],
		'field_3' => 'field 3'
	]
];

$mergedArray = array_merge_recursive($sourceArray, $remoteArray);
echo '<pre>';
var_dump($mergedArray);
?>
```

returns:
```sh
array(1) {
  ["page_one"]=>
  array(3) {
    ["field_1"]=>
    string(7) "field 1"
    ["field_2"]=>
    array(2) {
      ["sub_1"]=>
      string(5) "sub 1"
      ["sub_2"]=>
      array(2) {
        [0]=>
        string(5) "sub 2"
        [1]=>
        string(5) "sub 2"
      }
    }
    ["field_3"]=>
    string(7) "field 3"
  }
}
```

`sub_2` was duplicated.

Comment (BernardoSilva) [inline src/Utilities/HasUtilitiesTrait.php]: @antonvas don't remember the specific cases, but when I checked with @amyboyd it was having funny behaviour.
In first refactoring we end up not using it, now I've tried to use it, but got again some strange behaviour in deeper levels of arrays.

**Comment by antonvas** [comment]
Great :+1:

**Pull request merged**